### DISCUSSION TITLE
How to display SQL QUERY before execute it

### DISCUSSION THREAD

**Discussion by igormorgado** [collerek/ormar#456] [Q&A] [answered]
How to display the SQL Query before execute it, or somewhere between a sequence of filters.

My query is failing for KeyError, and I can't find exactly where.

**Comment by collerek** [comment, marked as answer]
Right now it's not possible in a clean way (unless `databases` have debug option, didn't check).
I will probably introduce one myself as for now in a need like this I revert to ugly printed 🤣 

If you execute a select you can uncomment line 277 in `queryset.py`:
![image](https://user-images.githubusercontent.com/16324238/144894948-7f4f4db6-0a50-4a42-803e-db034a7dc514.png)

If you use prefetch_related same in prefetch_query line 506
![image](https://user-images.githubusercontent.com/16324238/144895058-b7bdd828-b7ba-4950-ad5c-cd3b21a49fd0.png)

**Comment by mciszczon** [comment]
In Django there's a very helpful `.query` attribute on a QuerySet call, e.g.:

```python
print(Model.objects.filter(field="value").query)
```

would print something like:

```sql
SELECT * FROM `model` WHERE field ="value";
```

Having the possibility to have something like that in `ormar` would be awesome! Although I'm not sure where to start... As far as I understand, `ormar` itself is not responsible for transalting the ORM to sql-native code, but it actually is done by `databases` and the `sqlalchemy` dependency of it, right?

**Reply by collerek** [reply]
Yes, ormar uses sqlalchemy core, it's a kind of low-level SQL API (lower than orm that is).
Usually, you can call `build_select_expression()` to get the sqlalchemy query which you can print (with literal_binds to get parameters), although this query can change if you provide i.e. kwargs to last call (i.e. `get(name='Tom')` is equivalent to `filter(name='Tom').get()`) so there is no way to get the final query as the last operation is already async and executing the query.

In the so-called meantime, I checked and you can set the logging level to debug to get the queries executed from `databases`.

Logging is implemented in composite pks branch but it's a big change and I don't have much time to work on this as of now.
